From e267d333e28dacf128aa5a839e9e1252555959a5 Mon Sep 17 00:00:00 2001
From: noctis <970308389@qq.com>
Date: Mon, 3 Aug 2026 13:23:42 +0800
Subject: [PATCH 2/2] rustc_llvm: support Android LLVM 21 APIs

Android's LLVM 21 snapshot retains the LLVM 20 signatures for several target, lint, and ThinLTO interfaces. Keep using those interfaces until LLVM 22 while retaining the other LLVM 21-specific compatibility paths.

Change-Id: I5adf37ade0a5acfbae360df5eefc07e0fe9c42f0
---
 compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp | 8 ++++----
 compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp | 2 +-
 2 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp b/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp
index d4a05fbbbc..ea1417f1ac 100644
--- a/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp
+++ b/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp
@@ -512,7 +512,7 @@ extern "C" LLVMTargetMachineRef LLVMRustCreateTargetMachine(
 #endif
   }
 
-#if LLVM_VERSION_GE(21, 0)
+#if LLVM_VERSION_GE(22, 0)
   TargetMachine *TM = TheTarget->createTargetMachine(Trip, CPU, Feature,
                                                      Options, RM, CM, OptLevel);
 #else
@@ -834,7 +834,7 @@ extern "C" LLVMRustResult LLVMRustOptimize(
   if (LintIR) {
     PipelineStartEPCallbacks.push_back([](ModulePassManager &MPM,
                                           OptimizationLevel Level) {
-#if LLVM_VERSION_GE(21, 0)
+#if LLVM_VERSION_GE(22, 0)
       MPM.addPass(
           createModuleToFunctionPassAdaptor(LintPass(/*AbortOnError=*/true)));
 #else
@@ -1370,7 +1370,7 @@ LLVMRustCreateThinLTOData(LLVMRustThinLTOModule *modules, size_t num_modules,
   // Convert the preserved symbols set from string to GUID, this is then needed
   // for internalization.
   for (size_t i = 0; i < num_symbols; i++) {
-#if LLVM_VERSION_GE(21, 0)
+#if LLVM_VERSION_GE(22, 0)
     auto GUID =
         GlobalValue::getGUIDAssumingExternalLinkage(preserved_symbols[i]);
 #else
@@ -1693,7 +1693,7 @@ extern "C" void LLVMRustComputeLTOCacheKey(RustStringRef KeyOut,
 #endif
 
   // Based on the 'InProcessThinBackend' constructor in LLVM
-#if LLVM_VERSION_GE(21, 0)
+#if LLVM_VERSION_GE(22, 0)
   for (auto &Name : Data->Index.cfiFunctionDefs().symbols())
     CfiFunctionDefs.insert(GlobalValue::getGUIDAssumingExternalLinkage(
         GlobalValue::dropLLVMManglingEscape(Name)));
diff --git a/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp b/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp
index 72369ab7b6..5ef76191f4 100644
--- a/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp
+++ b/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp
@@ -153,7 +153,7 @@ extern "C" LLVMContextRef LLVMRustContextCreate(bool shouldDiscardNames) {
 
 extern "C" void LLVMRustSetNormalizedTarget(LLVMModuleRef M,
                                             const char *Target) {
-#if LLVM_VERSION_GE(21, 0)
+#if LLVM_VERSION_GE(22, 0)
   unwrap(M)->setTargetTriple(Triple(Triple::normalize(Target)));
 #else
   unwrap(M)->setTargetTriple(Triple::normalize(Target));
-- 
2.43.0

